Improve perf of int4pack_mm metal shader

manuelcandales opened the pull request:
Summary:
Improved perf by 5x

Using https://github.com/malfet/llm_experiments for benchmarking on Apple M1 Pro

Test Plan: CI

Differential Revision: D57731779

swolchok commented on aten/src/ATen/native/mps/operations/Quantized.mm:
```diff
@@ -22,41 +22,89 @@
+
+    using vecT = typename Vec4Type<T>::type;
+    constant vecT *A_ptr = reinterpret_cast<constant vecT *>(A + m * 4 * K);
+    //constant uchar2 *B_ptr = reinterpret_cast<constant uchar2 *>(B + (nb * 16 * K));
```
delete?

swolchok commented on aten/src/ATen/native/mps/operations/Quantized.mm:
```diff
@@ -22,41 +22,89 @@
-        uchar b_val = B_ptr[(k * ldb + (n % 32))/2];
-        b_val = (n & 1) == 0 ? b_val & 0x0f : (b_val >> 4);
-        rc += a_val * float(scale * T(b_val) + zero);
+      const T scale0 = scalesAndZeros[(kb * N + 4 * n) * 2 + 0];
+      const T zero0 = scalesAndZeros[(kb * N + 4 * n) * 2 + 1] - scale0 * T(8);
+
+      const T scale1 = scalesAndZeros[(kb * N + 4 * n + 1) * 2 + 0];
+      const T zero1 = scalesAndZeros[(kb * N + 4 * n + 1) * 2 + 1] - scale1 * T(8);
+
+      const T scale2 = scalesAndZeros[(kb * N + 4 * n + 2) * 2 + 0];
+      const T zero2 = scalesAndZeros[(kb * N + 4 * n + 2) * 2 + 1] - scale2 * T(8);
+
+      const T scale3 = scalesAndZeros[(kb * N + 4 * n + 3) * 2 + 0];
+      const T zero3 = scalesAndZeros[(kb * N + 4 * n + 3) * 2 + 1] - scale3 * T(8);
+
+      const float4 scales = float4(scale0, scale1, scale2, scale3);
+      const float4 zeros = float4(zero0, zero1, zero2, zero3);
```
nit: can we use a loop over `[0, 4)` instead? should get unrolled just fine

swolchok commented on aten/src/ATen/native/mps/operations/Quantized.mm:
```diff
@@ -22,41 +22,89 @@
+      const float4 scales = float4(scale0, scale1, scale2, scale3);
+      const float4 zeros = float4(zero0, zero1, zero2, zero3);
+
+      for(uint idx = 0; idx < groupSize && k < K; idx += 4, k += 4) {
```
what if `K % 4 != 0`? I don't see a check for that

manuelcandales replied:
K is supposed to be a multiple of 4. Actually, it should be a multiple of 32, based on the TORCH_CHECKS here:
https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/LinearAlgebra.cpp?fbclid=IwAR3k1YtoBlM8b0H2AkMfdPr10Yr4nIhHHVFgXEp1o9-TmeSDk39ZM58lrUw#L3439

swolchok commented:
> Using https://github.com/malfet/llm_experiments for benchmarking on Apple M1 Pro

what command did you run? what was the result?

swolchok commented on aten/src/ATen/native/mps/operations/Quantized.mm:
```diff
@@ -22,41 +22,89 @@
+      const float4 zeros = float4(zero0, zero1, zero2, zero3);
+
+      for(uint idx = 0; idx < groupSize && k < K; idx += 4, k += 4) {
+        float4x4 a_mat;
+        for(int j = 0; j < 4; ++j) {
+          a_mat[j] = float4(A_ptr[k/4 + j * K / 4]);
+        }
+
+        float4x4 t_b_mat;
+        for(int j = 0; j < 4; ++j) {
+          uchar b_val0 = B_ptr[((k + j) * ldb + ((4 * n) % 32))/2];
+          uchar b_val1 = B_ptr[((k + j) * ldb + ((4 * n) % 32))/2 + 1];
```
also, I am confused. this is a matrix-matrix kernel, right? I thought the majority use case in LLMs was matrix-vector

manuelcandales committed:
Improve perf of int4pack_mm metal shader (#126985)

Summary:

Improved perf by 5x

Using https://github.com/malfet/llm_experiments for benchmarking on Apple M1 Pro

Test Plan: CI

Reviewed By: malfet

Differential Revision: D57731779
